### PR TITLE
revset: add `fork_point` function

### PR DESCRIPTION
**Pull request by bnjmnt4n**
This can be used to find the best common ancestors of a revset with an arbitrary number of commits, which cannot be done currently.

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [X] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [X] I have added tests to cover my changes

### CODE REVIEW

**Comment by algmyr** [comment]
I don't think GCA is the term used for this, it's [lowest (or sometimes last) common ancestor](https://en.wikipedia.org/wiki/Lowest_common_ancestor). I would probably be fine with `lca`, but maybe others might not be as used to the abbreviations. Expanding it fully feels quite wordy though. Maybe `common_ancestor(revset)` is precise/descriptive enough?

File: lib/src/default_index/revset_engine.rs [hunk 1]
```diff
@@ -957,6 +957,21 @@ impl EvaluationContext<'_> {
...
+                    return Ok(Box::new(EagerRevset::empty()));
+                };
+                let mut positions = vec![position?];
+                for position in expression_positions_iter {
+                    positions = index
+                        .common_ancestors_pos(&positions, [position?].as_slice())
+                        .into_iter()
+                        .collect_vec();
+                }
+                Ok(Box::new(EagerRevset { positions }))
```
Comment (yuja) [inline lib/src/default_index/revset_engine.rs:973]: Perhaps, the order is reversed because `common_ancestors_pos()` returns `BTreeSet`?

File: lib/src/revset.rs [hunk 2]
```diff
@@ -253,6 +253,7 @@ pub enum RevsetExpression<St: ExpressionState> {
         candidates: Rc<Self>,
         count: usize,
     },
+    GreatestCommonAncestors(Rc<Self>),
```
Comment (yuja) [inline lib/src/revset.rs:256]: nit: I would move it closer to heads/roots because "gca" is DAG-based operation.

**Review by yuja — COMMENTED** [review]
Perhaps, it's the "greatest" in terms of generation numbers? I feel both `gca` and `lca` are equally cryptic.

`common_ancestor()`, `best_ancestor()`, `closest_ancestor()`?

**Comment by bnjmnt4n** [comment]
Git calls it best common ancestor (via [`git merge-base`](https://git-scm.com/docs/git-merge-base)):

> git merge-base finds the best common ancestor(s) between two commits to use in a three-way merge. One common ancestor is better than another common ancestor if the latter is an ancestor of the former. A common ancestor that does not have any better common ancestor is a best common ancestor, i.e. a merge base. Note that there can be more than one merge base for a pair of commits.

File: lib/src/default_index/revset_engine.rs (hunk 1 above)
Comment (bnjmnt4n) [reply lib/src/default_index/revset_engine.rs:973]: Ah, because all revsets must return items in reverse index position?

**Comment by bnjmnt4n** [comment]
Does anyone else have any thoughts on the naming? Also, should the name be in singular or plural form (`common_ancestor` vs `common_ancestors`)? I went with plural because the function can return multiple results if there are multiple common ancestors.

**Comment by yuja** [comment]
If we don't say "best", "greatest", etc., a singular form would suggest that it won't include all common ancestors. So I prefer singular `common_ancestor()`.

**Comment by bnjmnt4n** [comment]
Now that I think about it, I agree that `common_ancestors` seems to indicate `::a & ::b & ::c & ...`. What about `best_common_ancestors`?

**Comment by yuja** [comment]
> What about `best_common_ancestors`?

That seems good, but isn't it too wordy? I'm okay with either `common_ancestor()` (singular), `best_common_ancestor[s]()` (singular/plural). The return value is usually singular.

**Comment by emilazy** [comment]
Would `ancestor` be too much? I feel the fact that it’s common is implicit to the fact that you’re passing multiple arguments. Also, I might be misreading, but `ancestors()` should be `root()` (it looks like it might currently be `none()`?).

**Comment by bnjmnt4n** [comment]
> Would `ancestor` be too much? I feel the fact that it’s common is implicit to the fact that you’re passing multiple arguments.

`ancestor` is way to close to `ancestors`, which obtains the ancestors of all commits in the given revset. Additionally, both functions technically accept only a single revset argument, but the revset can resolve to multiple commits instead.

> Also, I might be misreading, but `ancestor()` should be `root()` (it looks like it might currently be `none()`?).

This is an interesting point, I'm not familiar enough with graph theory (?) to say if the best common ancestors of an empty set of nodes is the root of the graph.

**Comment by algmyr** [comment]
I think the best common ancestor of a single commit should be itself, and the best common ancestor of an empty set ought to be none (it doesn't exist)

**Comment by emilazy** [comment]
Hmm, you might actually be right. We want the property that `ancestor(A | B)` is `ancestor(ancestor(A) | ancestor(B))`, right? Therefore, yes, `ancestor(A)` should be `A`, and `ancestor(none())` should be the identity element, and I guess `none()` is that identity?

**Comment by martinvonz** [comment]
Something like `fork_point()` might also work. I'm not sure I like it better, but figured I'd share it at least.

**Comment by martinvonz** [comment]
> How can a commit be its own ancestor?

We simply define it such that a commits ancestor includes itself. We do that because it's often useful in practice. Mercurial does the same.

**Comment by emilazy** [comment]
Or, to motivate it directly in terms of a defining property: the greatest common ancestor of a set of commits is the commit that all elements of the set are descendents of, such that there is no descendent of the ancestor that also satisfies that property. For a single commit, that’s the commit itself. (This still leaves the choice of the ancestor of an empty set somewhat arbitrary, though it already makes it clear that there can be no single unique choice for many commit graphs, which motivates `none()` as a result.)

**Comment by joyously** [comment]
I think if it includes itself, it shouldn't be called ancestor. Commits don't appear out of thin air, do they? Wouldn't there at least be `root()`?

**Comment by algmyr** [comment]
> I think if it includes itself, it shouldn't be called ancestor.

In graph theory, when talking about things like [lowest common ancestor (LCA)](https://en.wikipedia.org/wiki/Lowest_common_ancestor) you consider nodes to be descendants of themselves.

**Comment by emilazy** [comment]
It’s also the same case as the greatest common divisor of a single number being itself (which Python’s `math.gcd` implements).

**Comment by joyously** [comment]
> In graph theory, when talking about things like [lowest common ancestor (LCA)](https://en.wikipedia.org/wiki/Lowest_common_ancestor) you consider nodes to be descendants of themselves.

If this is simply for the name of a function in code, call it whatever and explain with a comment.
If it's something a user will see, don't confuse us with "you are your own ancestor", even if another VCS did it.

**Comment by emilazy** [comment]
It’s the standard graph theory terminology, not a VCS‐specific quirk, and the mathematically obvious definition. People anyway aren’t likely to call `ancestor` on single commits except as a result of computed revsets, where the single‐commit case will provide correct results that are consistent with its behaviour on revsets of other sizes.

**Comment by algmyr** [comment]
> Hmm, you might actually be right. We want the property that `ancestor(As | Bs)` is `ancestor(ancestor(As) | ancestor(Bs))`, right? Therefore, yes, `ancestor(A)` (where `A` is a single commit) should be `A`, and `ancestor(none())` should be the identity element, and I guess `none()` is that identity?

This is mostly a tangent, but the empty case is a bit iffy and you end up having to be careful in how you define what a common ancestor is (to avoid issues of vacuous truths).

E.g. for a set of nodes `S` and we want to define the set of all common ancestors (ignoring the lowest/best part), it's tempting to try

"`common_ancestors(S)` are nodes with all nodes in `S` as descendants"

But this behaves poorly if `S` is empty. It's vacuously true that any node is an ancestor of all nodes in the empty set!

If you make the (sensible) constraint that common ancestors need to be an ancestors of some node in `S`, then you should end with nicer behavior that gives you the `none` result.

**Comment by emilazy** [comment]
Some more motivation of the base cases here:

If we desire that `ancestor(as | bs)` equals `ancestor(ancestor(as) | ancestor(bs))`, as it does for the >1 commit case, then we immediately get `ancestor(as)` = `ancestor(as | none())` (because `as | none()` = `none()`) = `ancestor(ancestor(none()) | ancestor(bs))`, which can only (as far as I can see) be equal to `ancestor(none() | bs)` if `ancestor(none())` = `none()` and `ancestor(ancestor(bs))` = `ancestor(bs)`, which gives us both the zero and one commit cases (assuming I understand correctly that `ancestor()` should never return a revset of more than one commit – although I wonder if that’s actually a desirable property, as I think it differs from `heads(::a & ::b)`?).

More pragmatically, a common practical use of this operation – Git’s `git merge-base` – is to answer the question “what’s the most ‘recent’ commit I can base my work on if I need it to be able to cleanly merge into these branches without pulling in additional commits only present in some of them?”. For the single‐commit case, the answer is that you can base that work on the commit itself, since there’s nothing “missing” from the non‐existent other branches. (Although, in that case I think it doesn’t give us a clear answer for `ancestor(none())`, since I think `root()` would be a valid answer there; it’s other properties that rule that case out.)

Base cases are hard. It wasn’t that long ago that people found the idea of zero very alien, and rejected formulas like *n* + 0 = n, *n* × 0 = 0, or the idea that you could sum up no integers at all and get 0, or take the product of no integers at all and get 1. But once you work out the rules it’s clear that there’s only one correct answer, and that skipping defining it because it seems weird and unintuitive at first reduces the power and convenience of your operations and forces you to introduce special cases elsewhere in the system.

**Comment by yuja** [comment]
> assuming I understand correctly that `ancestor()` should never return a revset of more than one commit

There's an edge case that `ancestor()` is resolved to multiple revisions, so I'm afraid that we cannot define `ancestor(*xs)` based on `ancestor(a | b | ...)`.
https://github.com/martinvonz/jj/pull/4795/commits/a1ced5b280f2f0af1c00261f5998faba51340981#diff-6bd5d4c0fd459e2c1cb3f1e0d880d6f23c4e3b11863adc76db3931e4b94d194dR2524-R2542

> Something like `fork_point()` might also work.

Sounds also good. It would reflect user's intent.

**Comment by martinvonz** [comment]
> > Something like `fork_point()` might also work.
> 
> Sounds also good. It would reflect user's intent.

Actually, maybe even just `forks()` since we have `merges()` (not `merge_point()`). "Forks" and "merges" seem to work well as analogies to roads forking and merging.

**Comment by yuja** [comment]
> EDIT: But an important difference between them is that `merges()` doesn't take an input set, so maybe `forks()` (and `fork_points()`?) would make it sound like `forks()` returns all commits with more than one child.

Yes, maybe we can add `merge_point(xs)` that finds the best common descendants, but I don't know if this name makes sense.

**Comment by martinvonz** [comment]
There doesn't seem to be a clear consensus and no one seems to feel strongly. To avoid this just getting stalled, I'm fine with @bnjmnt4n making the final decision on the name. Sounds good? Or do other feel like there's a name that's clearly best (and or names that they really don't want)?

**Comment by bnjmnt4n** [comment]
I like the suggested `fork_point`, which is relatively short and doesn't have any immediately ambiguous/conflicting meanings. I'm not too sure about the preferred definition of `fork_point` in `docs/revsets.md`: would anyone have any suggestions for the definition?

**Comment by algmyr** [comment]
Maybe it makes sense to give the example of the two commit case (which I think is `heads(::x & ::y)`) and just say fork_points generalizes this to an arbitrary number of commits?

**Comment by yuja** [comment]
We have "examples" section in the doc. `fork_point()` should probably be added there.

File: docs/revsets.md [hunk 3]
```diff
@@ -362,6 +368,18 @@ given [string pattern](#string-patterns).
...
+
+    * `fork_point(E|D)` ⇒ `{A}`
+    * `fork_point(E|C)` ⇒ `{A}`
+    * `fork_point(E|B)` ⇒ `{B}`
+    * `fork_point(E|A)` ⇒ `{A}`
+    * `fork_point(D|C)` ⇒ `{C}`
+    * `fork_point(D|B)` ⇒ `{A}`
+    * `fork_point(B|C)` ⇒ `{A}`
+    * `fork_point(A)` ⇒ `{A}`
+    * `fork_point()` ⇒ `{}`
```
Comment (martinvonz) [inline docs/revsets.md:381]: nit: maybe `fork_point(none())` so users don't think it can be called with no arguments

**Review by martinvonz — APPROVED** [review]

**Comment by arxanas** [comment]
"Fork point" already has a specific meaning in Git that involves traversing both the commit ancestry and the reflog. Is it intentional to have adopted the same term here? I was quite confused by the request in #4915 referencing the "fork point" because I didn't see why "three-dot diffs" would use (Git) fork points.

**Comment by arxanas** [comment]
> We talked about possibly using merge_point() for the opposite here: https://github.com/martinvonz/jj/pull/4795#issuecomment-2475119832

That makes sense to me.

---

Since I'm already here, here's some other thoughts, but I'm not specifically proposing anything:

- [not in favor] I agree that all the terms in the family "nearest/closest/least/greatest/best" common ancestor are unintuitive and wordy.
- [not in favor] I didn't see mentioned in this thread the terms from order theory: https://en.wikipedia.org/wiki/Join_and_meet
  - "join" and "meet" are quite concise and probably used extensively in the field, but I could never remember which was which. The names are not really intuitive, so I would not use them here.
  - "supremum" and "infimum" are... actually not the worst, in my opinion, because many users can leverage their existing etymological intuition to remember what they mean. But they're not "real" words outside of the mathematical sphere, and they're slightly longer, so I could easily see people failing to remember them, or how to spell them. (`sup` and `inf` are real abbreviations, as I recall, and I think would be better than e.g. `gca`, but still not ideal.)
  - "least upper bound" and "greatest lower bound" *sound* like they have the same problem as "least/greatest common ancestor", but I've seen them used much more consistently and much less confusingly. (`lub`/`glb` are also usable abbreviations and sometimes used in other contexts, like type systems, but worse than `sup`/`inf` in my opinion.) They're still fairly wordy and, while more intuitive than "join"/"meet", maybe still not sufficiently intuitive for users?
- [neutral] I think "fork" is a fine term overall, particularly as it has a natural analogue in trees, etc., similar to the "branch" metaphor. It just has historical baggage and ambiguity:
  - The meaning of "fork point" specifically as in Git is confusing, and, honestly, I question whether people really use this feature much. I think noting that our `fork_point` is unrelated to that "fork point" in the docs would be fine.
  - "fork" itself has a specific meaning in OSS, and "forks" in the sense that we're now describing them *sometimes* agree with the OSS "fork" meaning, which is unfortunate because some users might only think they understand what it means. I'm not convinced that we can't overcome this problem in documentation too, though.
  - One could argue that the natural dual to "fork" is ["join"](https://en.wikipedia.org/wiki/Fork%E2%80%93join_model) rather than "merge".
- [not in favor] We could also work backwards and ask "what's the opposite of a 'merge'?".
  - Maybe it would be a "split" point or a "divide"/"division" point? "Diverge"/"divergence" point? They're not the worst terms, in my opinion, but probably don't offer a significant advantage over "fork".
  - In the tree metaphor, besides the term "fork", I'm sure some people have called it a "branch" point or a "branching" point, but of course it seems dangerous to use "branch" as a term due to the confusion with specifically Git-style branches.
- [question] Is there a reason not to adopt the existing terminology of "merge-base" from Git?
  - It's not necessarily *obvious*,  but it seems precise, and probably has a reasonable degree of penetration as a widespread term.
  - We could also shorten it to "base"? It's a little more concise, but still feels fairly precise and descriptive. I could see people casually discussing "what's the base commit for this?" and have it be both intuitive and technically correct, which I think are strong factors in favor of the term.

**Comment by emilazy** [comment]
I was actually thinking of proposing `base()` for the bikeshed, so consider this a seconding of that proposal :) I can see people confusing it with `roots()`, but that seems okay; there’s probably no terminological choice that won’t seem ambiguous in some contexts.

**Comment by martinvonz** [comment]
> One could argue that the natural dual to "fork" is ["join"](https://en.wikipedia.org/wiki/Fork%E2%80%93join_model) rather than "merge".

When talking about road, I think it's "fork" and "merge".
